waspc/src/Wasp/Analyzer/Parser/Parser.hs

```diff
@@ -0,0 +1,309 @@
+{-# LANGUAGE LambdaCase #-}
+
+module Wasp.Analyzer.Parser.Parser
+  ( -- * AST to CST conversion
+
+    -- | This module takes @["SyntaxNode"]@ produced by 'Wasp.Analyzer.Parser.ConcreteParser'
+    -- and converts it into an abstract syntax tree.
+    parseStatements,
+    parseExpression,
+  )
+where
+
+import Control.Arrow (Arrow (first))
+import Control.Monad.Except (Except, runExcept, throwError)
+import Control.Monad.State.Strict (StateT, evalStateT, gets, modify)
+import Data.Maybe (catMaybes)
+import Wasp.Analyzer.Parser.AST (AST, Expr, ExtImportName, Identifier, Stmt)
+import qualified Wasp.Analyzer.Parser.AST as AST
+import Wasp.Analyzer.Parser.ConcreteParser.CST (SyntaxKind, SyntaxNode (SyntaxNode))
+import qualified Wasp.Analyzer.Parser.ConcreteParser.CST as S
+import Wasp.Analyzer.Parser.Ctx (Ctx (Ctx), WithCtx (WithCtx), ctxFromRgn)
+import Wasp.Analyzer.Parser.ParseError
+import Wasp.Analyzer.Parser.SourcePosition (SourcePosition (SourcePosition))
+import Wasp.Analyzer.Parser.SourceRegion (SourceRegion (SourceRegion))
+import qualified Wasp.Analyzer.Parser.Token as T
+
+data ParseState = ParseState
+  { pstateLine :: !Int,
+    pstateColumn :: !Int,
+    pstateRemainingSource :: String
+  }
+
+pstatePos :: ParseState -> SourcePosition
+pstatePos s = SourcePosition (pstateLine s) (pstateColumn s)
+
+type ParserM a = StateT ParseState (Except ParseError) a
+
+-- | An operation that turns some syntax nodes into a value and outputs the
+-- remaining nodes.
+type Action a = [SyntaxNode] -> ParserM (a, [SyntaxNode])
+
+-- | @parseStatements sourceString syntax@ tries to convert a concrete syntax
+-- tree into an AST.
+parseStatements :: String -> [SyntaxNode] -> Either ParseError AST
+parseStatements source syntax = runExcept $ evalStateT (coerceProgram syntax) initialState
+  where
+    initialState =
+      ParseState
+        { pstateLine = 1,
+          pstateColumn = 1,
+          pstateRemainingSource = source
+        }
+
+-- | @parseExpression sourceString syntax@ tries to convert a concrete syntax
+-- tree into an AST representing a single expression. Currently, it allows extra
+-- syntax after the single expression.
```

Comment:
It means that it ignores all syntax nodes after the expression ends. For example, if you gave it a list of nodes like `[IntLiteral, Whitespace, Whitespace, StringLiteral]` it would make an AST for just that `IntLiteral`. This shouldn't have any problems in practice though, since if you use `parseCSTExpression` there will never be extra syntax nodes after the expression (except for trivia tokens).

Reply:
Could you maybe mention this in the comment? Better to comment a bit too much here than too little.